openhtf/io/output/upload_json.py

```diff
@@ -0,0 +1,72 @@
+"""Module for uploading JSON-formatted test record files to server."""
+
+import os
+import logging
+import requests
+
+from json import JSONEncoder
+from openhtf import conf
+from openhtf.io.output.json_factory import OutputToJSON
+
+conf.Declare('project', 'project name')
+conf.Declare('data_server', 'data server')
+_LOG = logging.getLogger(__name__)
+
+class UploadJSON(OutputToJSON):
+  """Return an output callback that upload JSON Test Records to data server.
+
+  An example filename_pattern might be:
+    '/data/test_records/%(dut_id)s.%(start_time_millis)s'
+
+  To use this output mechanism:
+    test = openhtf.Test(PhaseOne, PhaseTwo)
+    test.AddOutputCallback(UploadJson(
+        '/data/test_records/%(dut_id)s.%(start_time_millis)s'))
+  
+    in config.yaml:
+       data_server:  http://your_data_server_entry
+  Args:
+    filename_pattern: A format string specifying the filename to write to,
+      will be formatted with the Test Record as a dictionary.
+    inline_attachments: Whether attachments should be included inline in the
+      output.  Set to False if you expect to have large binary attachments.
+  """
+
+  def __init__(self, filename_pattern=None, inline_attachments=True, **kwargs):
+    super(UploadJSON, self).__init__(filename_pattern, inline_attachments, **kwargs)
```

Comment:
filename_pattern defaults to None for OutputToJSON so that the encode() method can be called without writing to a file, but in order to invoke __call__, it *must* be set (you'll get an AssertionError otherwise).

If there's no filename_pattern passed in to your constructor as a fallback mechanism, you should set one using tempfile to generate a temporary file or directory to use instead and pass that to the super() constructor.